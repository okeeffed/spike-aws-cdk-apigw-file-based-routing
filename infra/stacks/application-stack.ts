import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
// import { ServerlessAuroraConstruct } from "./aurora-serverless-construct";
// import * as ec2 from "aws-cdk-lib/aws-ec2";
// import { DrizzleLambdaLayerConstruct } from "./drizzle-lambda-layer-contruct";
import { FileBasedApiGwConstruct } from "../lib/file-based-apigw-construct";
import path = require("path");
import { LlrtFunction } from "cdk-lambda-llrt";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Queue } from "aws-cdk-lib/aws-sqs";
import { SqsWithDlqCwConstruct } from "../lib/sqs-with-dlq-cw-lambda-construct";
import { SqsJobsConstruct } from "../lib/sqs-jobs-construct";
import { LambdaData } from "../types";
import * as logs from "aws-cdk-lib/aws-logs";

interface ApplicationStackProps extends cdk.StackProps {
  lambdasData: LambdaData[];
  lambdaLogGroups: Map<string, logs.LogGroup>;
}

export class ApplicationStack extends cdk.Stack {
  public fileBasedApiGw: FileBasedApiGwConstruct;

  constructor(scope: Construct, id: string, props: ApplicationStackProps) {
    super(scope, id, props);

    // Creating the queues for jobs
    const jobQueues = new SqsWithDlqCwConstruct(this, "ExampleQueue", {
      jobsDirectoryPath: path.join(__dirname, "../../dist/jobs"),
    });

    const exampleQueue = jobQueues.queueMap.get("example-queue") as Queue;

    this.fileBasedApiGw = new FileBasedApiGwConstruct(
      this,
      "FileBasedApiGw-Test",
      {
        restApiProps: {
          restApiName: "file-based-api-gw-test",
        },
        lambdaProps: {
          // TODO: When testing out RDS
          // layers: [drizzleLambdaLayer.layer],
          // TODO: When completed
          // role: auroraServerless.lambdaExecutionRole,
          // vpc,
          environment: {
            // DB_CONNECTION_URL: dbConnectionUrl, // The constructed URL
            // DB_SECRET_ARN: auroraServerless.cluster.secret?.secretArn || "", // The ARN of the secret containing the credentials

            // Note: this will only be accessible by the Lambda function that has the correct permissions
            EXAMPLE_QUEUE_URL: exampleQueue.queueUrl,
          },
        },
        lambdaDirectoryPath: path.join(__dirname, "../../dist/api"),
        LambdaFunctionClass: process.env.LLRT ? LlrtFunction : NodejsFunction,
        lambdasData: props.lambdasData,
        lambdaLogGroups: props.lambdaLogGroups,
      }
    );

    const jobToLambda = {
      "example-queue": this.fileBasedApiGw.lambdasMap.get("/v1/jobs-POST"),
    };

    for (const [queueName, lambda] of Object.entries(jobToLambda)) {
      if (!lambda) {
        throw new Error("Lambda not found for job");
      }

      const queue = jobQueues.queueMap.get(queueName);
      if (!queue) {
        throw new Error("Queue not found for job");
      }

      queue.grantSendMessages(lambda);
    }

    new SqsJobsConstruct(this, "SqsJobsConstruct", {
      jobsDirectoryPath: path.join(__dirname, "../../dist/jobs"),
      queueMap: jobQueues.queueMap,
    });
  }
}
