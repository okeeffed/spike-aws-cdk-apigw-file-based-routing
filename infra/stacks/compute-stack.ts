import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
// import { ServerlessAuroraConstruct } from "./aurora-serverless-construct";
// import * as ec2 from "aws-cdk-lib/aws-ec2";
// import { DrizzleLambdaLayerConstruct } from "./drizzle-lambda-layer-contruct";
import { FileBasedApiGwConstruct } from "../lib/file-based-apigw-construct";
import path = require("path");
import { LlrtFunction } from "cdk-lambda-llrt";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { LambdaData } from "../types";
import * as logs from "aws-cdk-lib/aws-logs";
import { Role } from "aws-cdk-lib/aws-iam";

interface ComputeStackProps extends cdk.StackProps {
  lambdasData: LambdaData[];
  lambdaLogGroups: Map<string, logs.LogGroup>;
  lambdaExecutionRole: Role;
}

export class ComputeStack extends cdk.Stack {
  public fileBasedApiGw: FileBasedApiGwConstruct;

  constructor(scope: Construct, id: string, props: ComputeStackProps) {
    super(scope, id, props);

    // Creating the queues for jobs
    // const jobQueues = new SqsWithDlqCwConstruct(this, "ExampleQueue", {
    //   jobsDirectoryPath: path.join(__dirname, "../../dist/jobs"),
    // });

    // const exampleQueue = jobQueues.queueMap.get("example-queue") as Queue;

    this.fileBasedApiGw = new FileBasedApiGwConstruct(
      this,
      "FileBasedApiGw-Test",
      {
        lambdaProps: {
          // TODO: When testing out RDS
          // layers: [drizzleLambdaLayer.layer],
          // TODO: When completed
          // role: auroraServerless.lambdaExecutionRole,
          // vpc,
          environment: {
            // DB_CONNECTION_URL: dbConnectionUrl, // The constructed URL
            // DB_SECRET_ARN: auroraServerless.cluster.secret?.secretArn || "", // The ARN of the secret containing the credentials
            // Note: this will only be accessible by the Lambda function that has the correct permissions
            // EXAMPLE_QUEUE_URL: exampleQueue.queueUrl,
            AWS_NODEJS_CONNECTION_REUSE_ENABLED: "1",
          },
          role: props.lambdaExecutionRole,
        },
        lambdaDirectoryPath: path.join(__dirname, "../../dist"),
        LambdaFunctionClass: process.env.LLRT ? LlrtFunction : NodejsFunction,
        lambdasData: props.lambdasData,
        lambdaLogGroups: props.lambdaLogGroups,
      }
    );
  }
}
