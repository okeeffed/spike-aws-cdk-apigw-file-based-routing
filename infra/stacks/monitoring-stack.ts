import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as logs from "aws-cdk-lib/aws-logs";
import { LambdaData } from "../types";

interface MonitoringStackProps extends cdk.StackProps {
  stage: string;
  lambdasData: LambdaData[];
}

function removeSpecialCharacters(input: string): string {
  return input.replace(
    // replace all special characters
    /[^a-zA-Z0-9]/g,
    ""
  );
}

export class MonitoringStack extends cdk.Stack {
  public logGroups: Map<string, logs.LogGroup> = new Map<
    string,
    logs.LogGroup
  >();
  constructor(scope: Construct, id: string, props: MonitoringStackProps) {
    super(scope, id, props);

    props.lambdasData.forEach((data) => {
      const functionName = removeSpecialCharacters(
        `${data.routePath}-${data.method}`
      );
      const logGroupName = `/aws/lambda/${props.stage}-${functionName}`;

      const logGroup = new logs.LogGroup(this, `${functionName}LogGroup`, {
        logGroupName: logGroupName,
        retention: logs.RetentionDays.ONE_WEEK,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });

      this.logGroups.set(data.entry, logGroup);
    });
  }
}
