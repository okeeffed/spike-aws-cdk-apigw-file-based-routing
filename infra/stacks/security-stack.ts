import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as iam from "aws-cdk-lib/aws-iam";
import { LambdaData } from "../types";

export interface SecurityStackProps extends cdk.StackProps {
  lambdasData: LambdaData[];
}

export class SecurityStack extends cdk.Stack {
  public lambdaExecutionRole: iam.Role;

  constructor(scope: Construct, id: string, props: SecurityStackProps) {
    super(scope, id, props);

    this.lambdaExecutionRole = new iam.Role(this, "LambdaExecutionRole", {
      assumedBy: new iam.ServicePrincipal("lambda.amazonaws.com"),
      description: "An execution role for Lambda functions",
    });

    // Step 2: Attach the AWSLambdaBasicExecutionRole policy to the role
    this.lambdaExecutionRole.addManagedPolicy(
      iam.ManagedPolicy.fromAwsManagedPolicyName(
        "service-role/AWSLambdaBasicExecutionRole"
      )
    );
  }
}
