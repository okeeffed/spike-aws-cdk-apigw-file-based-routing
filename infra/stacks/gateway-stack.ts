import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
// import { ServerlessAuroraConstruct } from "./aurora-serverless-construct";
// import * as ec2 from "aws-cdk-lib/aws-ec2";
// import { DrizzleLambdaLayerConstruct } from "./drizzle-lambda-layer-contruct";
import { LambdaData } from "../types";
import { ApplicationStack } from "./application-stack";
import { HttpApi, HttpMethod } from "aws-cdk-lib/aws-apigatewayv2";
import * as integrations from "aws-cdk-lib/aws-apigatewayv2-integrations";

interface ApplicationStackProps extends cdk.StackProps {
  lambdasData: LambdaData[];
  applicationStack: ApplicationStack;
}

export class GatewayStack extends cdk.Stack {
  public restApi: HttpApi;

  constructor(scope: Construct, id: string, props: ApplicationStackProps) {
    super(scope, id, props);

    // Instantiate an HTTP API
    const httpApi = new HttpApi(this, "MyHttpApi", {
      apiName: "file-based-api-gw-test",
    });

    for (const lambdaData of props.lambdasData) {
      const lambdaFn = props.applicationStack.fileBasedApiGw.lambdasMap.get(
        lambdaData.entry
      );

      if (!lambdaFn) {
        throw new Error(
          `Lambda function not found for entry: ${lambdaData.entry}`
        );
      }

      // Add routes to the HTTP API
      httpApi.addRoutes({
        path: lambdaData.routePath,
        methods: [
          HttpMethod[
            lambdaData.method.toUpperCase() as keyof typeof HttpMethod
          ],
        ],
        integration: new integrations.HttpLambdaIntegration(
          `${lambdaData.entry}Integration`,
          lambdaFn
        ),
      });
    }
  }
}
