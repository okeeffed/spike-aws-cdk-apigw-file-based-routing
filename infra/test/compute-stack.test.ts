import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { ComputeStack } from "../stacks/compute-stack";
import { SecurityStack } from "../stacks/security-stack";
import { MonitoringStack } from "../stacks/monitoring-stack";

describe("ComputeStack", () => {
  test("Gateway stack matches snapshot", () => {
    const lambdasData = [
      {
        entry: "/api/v1/posts/GET/index.js",
        routePath: "/v1/posts",
        method: "GET",
      },
      {
        entry: "/api/v1/posts/POST/index.js",
        routePath: "/v1/posts",
        method: "POST",
      },
    ];

    const app = new cdk.App();
    const iamStack = new SecurityStack(app, "SecurityStack");
    const monitoringStack = new MonitoringStack(app, "MonitoringStack", {
      lambdasData,
      stage: "dev",
    });

    const stack = new ComputeStack(app, "ComputeStack", {
      lambdasData,
      lambdaLogGroups: monitoringStack.logGroups,
      lambdaExecutionRole: iamStack.lambdaExecutionRole,
    });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
