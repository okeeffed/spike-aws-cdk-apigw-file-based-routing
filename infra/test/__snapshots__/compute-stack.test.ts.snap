// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComputeStack Gateway stack matches snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "FileBasedApiGwTestGETv1posts773712DA": {
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "d5dac4bae04a6a195a3edf9db5b42af3b328dcf2ba2890f07b6ee1844e09ab5c.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "LoggingConfig": {
          "LogGroup": {
            "Fn::ImportValue": "MonitoringStack:ExportsOutputRefv1postsGETLogGroup4A3A885AC51DA2CF",
          },
        },
        "Role": {
          "Fn::ImportValue": "SecurityStack:ExportsOutputFnGetAttLambdaExecutionRoleD5C26073ArnADAF6607",
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "FileBasedApiGwTestPOSTv1posts48438635": {
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "1a0c7919d7b6043d9d668f7e2a4797685a7864edc050d014cb15391d31e4600d.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "LoggingConfig": {
          "LogGroup": {
            "Fn::ImportValue": "MonitoringStack:ExportsOutputRefv1postsPOSTLogGroup90A1408C7143A1EE",
          },
        },
        "Role": {
          "Fn::ImportValue": "SecurityStack:ExportsOutputFnGetAttLambdaExecutionRoleD5C26073ArnADAF6607",
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
